core/domain/collection_services.py

```diff
@@ -247,6 +247,77 @@ def is_collection_summary_editable(collection_summary, user_id=None):
         or collection_summary.community_owned)
 
 
+def get_learner_collection_dict_by_id(
+        collection_id, user_id, strict=True, version=None):
+    """Creates and returns a dictionary representation of a collection given by
+    the provided collection ID. This dictionary contains extra information
+    along with the dict returned by collection_domain.Collection.to_dict()
+    which includes useful data for the collection learner view. The information
+    includes progress in the collection, information about explorations
+    referenced within the collection, and a slightly nicer data structure for
+    frontend work.
+    """
```

Comment:
Done. This also led me to fixing many bugs surrounding the deletion process for explorations and how they relate to collections. Now, if an exploration referenced within a collection is deleted, it will cause the collection player to issue an error. The collection editor now will properly display the collection node, but place a warning next to it to notify the creator that exploration is broken. The editor also now allows the deleted exploration to be safely removed from the collection, which then in turn fixes the collection player.

These errors then led me to wondering about public vs. private status of explorations and collections, so I added guards for those, as well. If a collection is private, it can include both private and public explorations. If it's public, it can only contain public explorations per the collections design document.

Finally, while investigating the above issue with rights, I also realized the collection editor is viewable by users without permission, even though they can't change anything. This has been changed; the collection editor will not allow users to view it who do not have permission to edit that collection, until a read-only version of the collection editor is implemented.